torch/csrc/distributed/autograd/engine/dist_engine.cpp

```diff
@@ -0,0 +1,268 @@
+      graphTask->outstanding_tasks_++;
+      queue.push(mapEntry.second.get());
+    }
+  }
+
+  edge_list recvBackwardEdges;
+  // Traverse the graph.
+  auto& dependencies = graphTask->dependencies_;
+  while (!queue.empty()) {
+    auto fn = queue.front();
+    queue.pop();
+
+    for (const auto& edge : fn->next_edges()) {
+      if (auto nextFn = edge.function.get()) {
+        dependencies[nextFn] += 1;
+        const bool wasInserted = seen.insert(nextFn).second;
+        if (wasInserted) {
+          // Seeing this function for the first time.
+          queue.push(nextFn);
+
+          if (nextFn->next_edges().empty()) {
+            TORCH_INTERNAL_ASSERT(
+                dynamic_cast<AccumulateGrad*>(nextFn) ||
+                dynamic_cast<RecvRpcBackward*>(nextFn));
+            // We have found a leaf node which should be either AccumulateGrad
+            // or RecvRpcBackward. Record the function
+            // to ensure we don't execute it and instead accumulate the grads on
+            // the autograd context. These functions would be passed in as the
+            // 'outputs' parameter of the vanilla autograd engine.
+
+            // We don't accumulate any grads in the context for RecvRpcBackward.
+            // RecvRpcBackward is added as an output edge to indicate it is a
+            // leaf node and this helps in properly computing dependencies for
+            // the local autograd graph. Putting RecvRpcBackward in
+            // 'outputEdges' means that this function needs to be executed
+            // (inline with our assumption for FAST mode that all send/recv
+            // functions are valid in the backward pass), and as a result all of
+            //  its ancestors need to be executed as well.
+            if (dynamic_cast<RecvRpcBackward*>(nextFn)) {
+              recvBackwardEdges.emplace_back(edge);
+            }
+            outputEdges.emplace_back(edge);
+          }
+        }
+      }
+    }
+  }
+
+  // Now lets compute which functions need to be executed. The algorithm is as
+  // follows:
+  // 1. Use 'init_to_execute' with the existing GraphRoot and outputEdges.
+  // 2. Create a dummy GraphRoot which points to all 'send' functions for this
+  //    context and run 'init_to_execute' again, but with empty 'outputs'
+  //    (since we already processed outputs in 1.). This ensures we mark
+  //    appropriate functions as needed if they are reachable only from a
+  //    specific 'send' function locally and not necessarily from the provided
+  //    roots.
+  // 3. For all edges in 'outputEdges' which point to 'RecvRpcBackward', mark
+  //    those functions as needed for execution. The reason for this is that
+  //    'init_to_execute', will mark these as not needed. But 'RecvRpcBackward'
+  //    is unique in the sense that we use it as a leaf node in graph to compute
+  //    needed execution accurately, but unlike AccumulateGrad, we do need to
+  //    execute this function.
```

Comment:
Hmm, this seems a bit complicated.

Reply:
OK, I think the algorithm is fine. However, I think this comment would be clearer if went something like "the set of nodes we must execute are", rather than couching it algorithmically